draft-ietf-quic-tls.md

```diff
@@ -808,26 +807,110 @@ packet number limit.  An endpoint MUST initiate a key update ({{key-update}})
 
 
-## Packet Number Protection {#pn-encrypt}
+## Header Protection {#header-protect}
 
-QUIC packet numbers are protected using a key that is derived from the current
-set of secrets.  The key derived using the "pn" label is used to protect the
-packet number from casual observation.  The packet number protection algorithm
-depends on the negotiated AEAD.
+Parts of QUIC packet headers, in particular the Packet Number field, are
+protected using a key that is derived separate to the packet protection key and
+IV.  The key derived using the "quic hp" label is used to provide
+confidentiality protection for those fields that are not exposed to on-path
+elements.
 
-Packet number protection is applied after packet protection is applied (see
-{{aead}}).  The ciphertext of the packet is sampled and used as input to an
-encryption algorithm.
+This protection applies to the least-significant bits of the first byte, plus
+the Packet Number field.  The four least-significant bits of the first byte are
+protected for packets with long headers; the five least significant bits of the
+first byte are protected for packets with short headers.  For both header forms,
+this covers the reserved bits and the Packet Number Length field; the Key Phase
+bit is also protected for packets with a short header.
 
-In sampling the packet ciphertext, the packet number length is assumed to be 4
-bytes (its maximum possible encoded length), unless there is insufficient space
-in the packet for sampling.  The sampled ciphertext starts after allowing for a
-4 byte packet number unless this would cause the sample to extend past the end
-of the packet.  If the sample would extend past the end of the packet, the end
-of the packet is sampled.
+This process does not apply to Retry or Version Negotiation packets, which do
+not contain a protected payload or any of the fields that are protected by this
+process.
 
-For example, the sampled ciphertext for a packet with a short header can be
-determined by:
+
+### Header Protection Application
+
+Header protection is applied after packet protection is applied (see {{aead}}).
+The ciphertext of the packet is sampled and used as input to an encryption
+algorithm.  The algorithm used depends on the negotiated AEAD.
+
+The output of this algorithm is 5 bytes of mask which are applied to the
+protected using exclusive OR.  The least significant bits of the first byte of
+the packet are masked by the first mask byte, and the packet number is masked
+with the remaining bytes.
+
+{{pseudo-hp}} shows a sample algorithm for applying header protection. Removing
+protection only differs in the order in which the packet number length
+(pn_length) is determined.
+
+~~~
+mask = header_protection(hp_key, sample)
+
+pn_length = (packet[0] & 0x03) + 1
+if packet[0] & 0x80 == 0x80:
+   # Long header: 4 bits masked
+   packet[0] ^= mask[0] & 0x0f
+else:
+   # Short header: 5 bits masked
+   packet[0] ^= mask[0] & 0x1f
```

Comment:
This doesn't say anything about peers that are not spinning, in which case it should be 6 bits.

Reply:
My understanding is that we cannot change which bits are protected based on the use (or non-use) is spin bits, because the use is not negotiated.

Having the bit always outside the header protection also gives the non-participating endpoints greater freedom to play with the bits (e.g. always set to zero).

Reply:
Why can't we? If I decide to not spin (in general, or just for this connection / CID), I can apply header protection to that bit as well. This would provide optimal greasing.

Reply:
To reconstruct the AAD that was used, the receiver needs to know if you have applied header protection to the spin bit. However, if you are using spin bit is never signaled to the peer.

Reply:
Unused spins are currently required to be a constant value randomly chosen to be 0 or 1. Not sure why, but encrypting it would violate that.

Reply:
One reason to choose a fixed value per connection is it makes it easier to detect which connections are not spinning.  Another reason is it's easy to leak information in that bit, so it's safer to just set a fixed value.

Reply:
@marten-seemann you can grease this bit without encrypting it - just send a different random value with every packet. I would prefer to keep the spec simpler by never encrypting this bit.

Reply:
@ianswett I think you can still tell which connections aren't spinning even if set to a random value. But I agree about leaking information unintentionally. 

I'd also like to make sure we grease this bit, but I would be fine with doing that in a way that doesn't involve encrypting it with the rest of the packet.

Reply:
@DavidSchinazi The simplest way to randomize it is to encrypt it, but as Ian says, there are reasons not to do this.

Reply:
No, that doesn't work, because the value of the spin bit is covered by the AAD, and we have to have peers agree on what the value is.